packages/xo-server/src/xapi/mixins/patching.js

```diff
@@ -464,4 +464,127 @@ export default {
+        }),
+        vm => vm.$resident_on?.$id
+      ),
+      vms =>
+        vms.map(vm => {
+          const mapVifsNetworks = {}
+          vm.$VIFs.forEach(vif => {
+            mapVifsNetworks[vif.$id] = vif.$network.$id
+          })
+
+          const mapVdisSrs = {}
+          flatMap(vm.$snapshots, '$VBDs')
+            .concat(vm.$VBDs)
+            .forEach(vbd => {
+              if (vbd.type === 'Disk') {
+                const vdi = vbd.$VDI
+                mapVdisSrs[vdi.$id] = vdi.$SR.$id
+              }
+            })
+
+          return {
+            id: vm.$id,
+            ref: vm.$ref,
+            mapVifsNetworks,
+            mapVdisSrs,
+          }
+        })
+    )
+
+    if (vmsByHost.undefined !== undefined) {
+      throw new Error('Could not find host of all running VMs')
+    }
+
+    // Cache $fields to make sure we can access them later
+    hosts = hosts.map(host => ({
+      ...host,
+      id: host.$id,
+      ref: host.$ref,
+      metricsId: host.$metrics.$id,
+      metricsRef: host.$metrics.$ref,
+    }))
+
+    // Put master in first position to restart it first
+    const indexOfMaster = hosts.findIndex(host => host.ref === this.pool.master)
+    if (indexOfMaster > 0) {
+      ;[hosts[0], hosts[indexOfMaster]] = [hosts[indexOfMaster], hosts[0]]
+    }
+
+    // Restart all the hosts one by one
+    for (const host of hosts) {
+      const hostId = host.id
+      const metricsId = host.metricsId
+
+      await this.barrier(host.metricsRef)
+      await this._waitObjectState(metricsId, metrics => metrics.live)
+      const rebootTime = parseDateTime(await this.call('host.get_servertime', host.ref))
+
+      log.debug(`Evacuate and restart host ${hostId}`)
+      await this.rebootHost(hostId)
+
+      log.debug(`Wait for host ${hostId} to be up`)
+      await Promise.race([
+        (async () => {
+          await this._waitObjectState(hostId, host => rebootTime < host.other_config.agent_start_time * 1e3)
+          await this._waitObjectState(metricsId, metrics => metrics.live)
+          await this._waitObjectState(hostId, host => host.enabled)
+        })(),
+        new Promise((resolve, reject) =>
+          setTimeout(() => reject(new Error(`Host ${hostId} took too long to restart`)), 5 * 60 * 1e3)
+        ),
+      ])
+      log.debug(`Host ${hostId} is up`)
+    }
+
+    log.debug('Migrate VMs back to where they were')
+
+    // Start with the last host since it's the emptiest one after the rolling
+    // update
+    ;[hosts[0], hosts[hosts.length - 1]] = [hosts[hosts.length - 1], hosts[0]]
+
+    let error
+    for (const host of hosts) {
+      const hostId = host.id
+      const vmsData = vmsByHost[hostId]
+
+      if (vmsData === undefined) {
+        continue
+      }
+
+      for (const vmData of vmsData) {
+        try {
+          await this.migrateVm(vmData.id, this, hostId, {
+            mapVifsNetworks: vmData.mapVifsNetworks,
+            mapVdisSrs: vmData.mapVdisSrs,
```

Comment:
Are we sure that the `VDI` IDs didn't change after their migration?

Reply:
As discussed: I'm reverting it to calling `migrateVm` without any extra param so that it triggers a `VM.pool_migrate` instead of `VM.migrate_send` because:
- apparently, XAPI won't allow an `evacuate` if there's a VDI on a local SR anyway, so the VDIs will be and remain on shared SRs during the evacuate
- when I'm migrating the VMs back using a `pool_migrate`, it won't try to migrate the VDIs to the default SR as I first thought, so the VDIs will again stay on their original (shared) SRs
- and regarding the networks, the `evacuate` will force the hosts to reconnect to the networks if necessary and a `pool_migrate` won't try to move the VIFs to the management PIF's network

So a `pool_migrate` should be enough in any situation.